pkg/core/transaction/witness_condition.go

```diff
@@ -0,0 +1,575 @@
+// deserialize condition-specific data.
+func (c *ConditionOr) DecodeBinarySpecific(r *io.BinReader, maxDepth int) {
+	a := readArrayOfConditions(r, maxDepth)
+	if r.Err == nil {
+		*c = a
+	}
+}
+
+// MarshalJSON implements json.Marshaler interface.
+func (c *ConditionOr) MarshalJSON() ([]byte, error) {
+	return arrayToJSON(c, []WitnessCondition(*c))
+}
+
+// Type implements WitnessCondition interface and returns condition type.
+func (c *ConditionScriptHash) Type() WitnessConditionType {
+	return WitnessScriptHash
+}
+
+// Match implements WitnessCondition interface checking whether this condition
+// matches given context.
+func (c *ConditionScriptHash) Match(ctx MatchContext) (bool, error) {
+	return util.Uint160(*c).Equals(ctx.GetCurrentScriptHash()), nil
+}
+
+// EncodeBinary implements WitnessCondition interface allowing to serialize condition.
+func (c *ConditionScriptHash) EncodeBinary(w *io.BinWriter) {
+	w.WriteB(byte(c.Type()))
+	w.WriteBytes(c[:])
+}
+
+// DecodeBinarySpecific implements WitnessCondition interface allowing to
+// deserialize condition-specific data.
+func (c *ConditionScriptHash) DecodeBinarySpecific(r *io.BinReader, _ int) {
+	r.ReadBytes(c[:])
+}
+
+// MarshalJSON implements json.Marshaler interface.
+func (c *ConditionScriptHash) MarshalJSON() ([]byte, error) {
+	aux := conditionAux{
+		Type: c.Type().String(),
+		Hash: (*util.Uint160)(c),
+	}
+	return json.Marshal(aux)
+}
+
+// Type implements WitnessCondition interface and returns condition type.
+func (c *ConditionGroup) Type() WitnessConditionType {
+	return WitnessGroup
+}
+
+// Match implements WitnessCondition interface checking whether this condition
+// matches given context.
+func (c *ConditionGroup) Match(ctx MatchContext) (bool, error) {
+	return ctx.CurrentScriptHasGroup((*keys.PublicKey)(c))
+}
+
+// EncodeBinary implements WitnessCondition interface allowing to serialize condition.
+func (c *ConditionGroup) EncodeBinary(w *io.BinWriter) {
+	w.WriteB(byte(c.Type()))
+	(*keys.PublicKey)(c).EncodeBinary(w)
+}
+
+// DecodeBinarySpecific implements WitnessCondition interface allowing to
+// deserialize condition-specific data.
+func (c *ConditionGroup) DecodeBinarySpecific(r *io.BinReader, _ int) {
+	(*keys.PublicKey)(c).DecodeBinary(r)
+}
+
+// MarshalJSON implements json.Marshaler interface.
+func (c *ConditionGroup) MarshalJSON() ([]byte, error) {
+	aux := conditionAux{
+		Type:  c.Type().String(),
+		Group: (*keys.PublicKey)(c),
+	}
+	return json.Marshal(aux)
+}
+
+// Type implements WitnessCondition interface and returns condition type.
+func (c ConditionCalledByEntry) Type() WitnessConditionType {
+	return WitnessCalledByEntry
+}
+
+// Match implements WitnessCondition interface checking whether this condition
+// matches given context.
+func (c ConditionCalledByEntry) Match(ctx MatchContext) (bool, error) {
+	entry := ctx.GetEntryScriptHash()
+	return entry.Equals(ctx.GetCallingScriptHash()) || entry.Equals(ctx.GetCurrentScriptHash()), nil
```

Comment:
I'm not sure, but is this condition the same as the one specified in C# and the one from ours `checkScope`:
```
if callingScriptHash.Equals(util.Uint160{}) || callingScriptHash == entryScriptHash {
					return true, nil
				}
```

Reply:
Effectively they're the same, but as we technically can't have `null` util.Uint160 we either have to compare to `util.Uint160{}` or current script (because entry == current in this case anyway).